# Transcriptome Annotation of 17 Porcine Tissues Using Nanopore Sequencing Technology
Jinghui Li, Dailu Guan, Michelle M. Halstead, Alma D. Islas-Trejo, Daniel E. Goszczynski, Catherine W. Ernst, Hao Cheng, Pablo Ross, Huaijun Zhou
## Project summary
The annotation of animal genomes plays an important role in elucidating molecular mechanisms behind the genetic control of economically important traits. Here, we employed the long-read sequencing technology, Oxford Nanopore Technology (ONT), to annotate the pig transcriptome across 17 tissues from two Yorkshire littermate pigs. More than 9.8 million reads were obtained from a single ONT flow cell, and 69,781 unique transcripts at 50,108 loci were identified. Of these transcripts, 16,255 were found to be novel isoforms, and 22,344 were found at novel loci compared to the Ensembl (release 102) and NCBI (release 106) references. As the same samples were used for cis-regulatory elements annotation in pig genome, the transcriptome annotation generated by this study provides an additional and complementary annotation resource for the Functional Annotation of Animal Genomes (FAANG) effort to comprehensively annotate the pig genome. 
## Public data
The Nanopore sequencing data generated in this study can be found in the Sequence Read Archive (SRA) database of the National Center for Biotechnology Information with the identifier PRJNA671673 (https://www.ncbi.nlm.nih.gov/bioproject/PRJNA671673). The code for data analyses is shown in scripts/nanopore_ann.smk, and the transcriptom annotation is in GTF/filterSamples.gtf
